vault/resource_database_secret_backend_role.go

```diff
@@ -20,121 +24,176 @@ var (
 				Optional:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Database statements to execute to renew a user.",
 			},
+			consts.FieldCredentialType: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Specifies the type of credential that will be generated for the role.",
+			},
+			consts.FieldCredentialConfig: {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Description: "Specifies the configuration for the given credential_type.",
+			},
+			consts.FieldKeyBits: {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  2048,
+				Description: "Number of bits to use for the generated keys. Options include: 2048 " +
+					"(default), 3072, 4096; with key_type=ec, allowed values are: 224, 256 (default), " +
+					"384, 521; ignored with key_type=ed25519.",
+			},
+			consts.FieldSignatureBits: {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  256,
+				Description: "The number of bits to use in the signature algorithm. Options include: 256 " +
+					"(default), 384, 512.",
+			},
+			consts.FieldFormat: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The output format of the generated private key credential.",
+			},
+
+			consts.FieldPasswordPolicy: {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "The policy used for password generation. Only used when credential type " +
+					"is 'password'.",
+			},
 		},
 	}
 }
 
-func databaseSecretBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
+func databaseSecretBackendRoleWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client, e := provider.GetClient(d, meta)
 	if e != nil {
-		return e
+		return diag.FromErr(e)
 	}
 
-	backend := d.Get("backend").(string)
-	name := d.Get("name").(string)
+	backend := d.Get(consts.FieldBackend).(string)
+	name := d.Get(consts.FieldName).(string)
 
 	path := databaseSecretBackendRolePath(backend, name)
 
 	data := map[string]interface{}{
-		"db_name":             d.Get("db_name"),
-		"creation_statements": d.Get("creation_statements"),
+		"db_name":             d.Get(consts.FieldDBName),
+		"creation_statements": d.Get(consts.FieldCreationStatements),
 	}
-
-	if v, ok := d.GetOkExists("default_ttl"); ok {
-		data["default_ttl"] = v
-	}
-	if v, ok := d.GetOkExists("max_ttl"); ok {
-		data["max_ttl"] = v
-	}
-	if v, ok := d.GetOkExists("revocation_statements"); ok && v != "" {
-		data["revocation_statements"] = v
+	fields := []string{
+		consts.FieldDefaultTTL,
+		consts.FieldMaxTTL,
+		consts.FieldRevocationStatements,
+		consts.FieldRollbackStatements,
+		consts.FieldRenewStatements,
+		consts.FieldCredentialType,
+		consts.FieldCredentialConfig,
+		consts.FieldCACert,
+		consts.FieldPrivateKey,
+		consts.FieldKeyType,
+		consts.FieldKeyBits,
+		consts.FieldSignatureBits,
+		consts.FieldCommonNameTemplate,
```

Comment:
I am not sure I understand why we are writing these fields to to Vault as top-level fields. Aren't these supposed to be nested under `credential_config`?

Reply:
Removed in the latest commits since they were not required

Reply:
Yeah, you are right. They are now removed since they are not top-level fields, and now we just write `credential_type` and `credential_config` to Vault.